### PR TITLE
SE-874-Tron-Inactive-Initiating-Action-Account-Level

### PR DESCRIPTION
**Pull request by Palmerg4**

### CODE REVIEW

File: src/client/application/ProtocolApplicationHandler.sol [hunk 1]
```diff
@@ -204,31 +207,90 @@ contract ProtocolApplicationHandler is
...
+                    ruleProcessor.checkAccountMaxValueByAccessLevel(accountMaxValueByAccessLevel[_action].ruleId, score, _balanceValuation, _transferValuation);
+                if (accountMaxValueOutByAccessLevel[ActionTypes.SELL].active) {
+                usdValueTotalWithrawals[_from] = ruleProcessor.checkAccountMaxValueOutByAccessLevel(
+                    accountMaxValueOutByAccessLevel[_action].ruleId,
+                    fromScore,
+                    usdValueTotalWithrawals[_from],
+                    _transferValuation
+                );
+            }
             }
```
Comment (Palmerg4) [inline src/client/application/ProtocolApplicationHandler.sol:233]: Format is slightly off here

File: test/client/token/ERC721/marketplace/BuyerMarketplaceTests.t.sol [hunk 2]
```diff
@@ -1199,5 +1199,218 @@ contract MarketplaceNonCustodialTestsErc20SellsNftBuys is TokenUtils, ERC721Util
...
+
+        switchToRiskAdmin();
+        applicationAppManager.addRiskScore(user1, riskScores[0]);
+        applicationAppManager.addRiskScore(user2, riskScores[1]);
+        
+        vm.startPrank(user1, user1);
+        vm.expectRevert(
+            abi.encodeWithSelector(
+                TransferFailed.selector, 
+                address(applicationNFTv2),
```
Comment (Palmerg4) [inline test/client/token/ERC721/marketplace/BuyerMarketplaceTests.t.sol:1226]: I may be wrong here, but shouldnt the error be emitted from applicationCoin, not applicationNFTv2?

File: src/client/application/ProtocolApplicationHandler.sol [hunk 3]
```diff
@@ -204,31 +207,90 @@ contract ProtocolApplicationHandler is
                 ruleProcessor.checkAccountDenyForNoAccessLevel(fromScore);
                 ruleProcessor.checkAccountDenyForNoAccessLevel(score);
             }
+            if (accountMaxValueByAccessLevel[_action].active && _to != address(0))
```
Comment (oscarsernarosero) [inline src/client/application/ProtocolApplicationHandler.sol:210]: Im not sure why we need to check here for `_to != address(0)` when we already know this is a P2P_TRANSFER

File: src/client/application/ProtocolApplicationHandler.sol [hunk 4]
```diff
@@ -204,31 +207,90 @@ contract ProtocolApplicationHandler is
...
+                    fromScore,
+                    usdValueTotalWithrawals[_from],
+                    _transferValuation
+                );
+            }
         } else if (_action == ActionTypes.BUY) {
-            if (isContract(_sender) && _from != _sender){ /// Non custodial buy
+            if (_from != _sender){ /// Non custodial buy
                 if (accountDenyForNoAccessLevel[ActionTypes.SELL].active) ruleProcessor.checkAccountDenyForNoAccessLevel(fromScore);
+                if (accountMaxValueByAccessLevel[ActionTypes.SELL].active && _to != address(0))
```
Comment (oscarsernarosero) [inline src/client/application/ProtocolApplicationHandler.sol:223]: same here

File: src/client/application/ProtocolApplicationHandler.sol [hunk 5]
```diff
@@ -204,31 +207,90 @@ contract ProtocolApplicationHandler is
...
+                    usdValueTotalWithrawals[_from] = ruleProcessor.checkAccountMaxValueOutByAccessLevel(
+                        accountMaxValueOutByAccessLevel[_action].ruleId,
+                        fromScore,
+                        usdValueTotalWithrawals[_from],
+                        _transferValuation
+                    );
+                }
             }
             if (accountDenyForNoAccessLevel[_action].active) ruleProcessor.checkAccountDenyForNoAccessLevel(score);
+            if (accountMaxValueByAccessLevel[_action].active && _to != address(0))
```
Comment (oscarsernarosero) [inline src/client/application/ProtocolApplicationHandler.sol:235]: and here

**Review by mpetersoCode55 — APPROVED** [review]
Approved for me! Great job man

File: src/client/application/ProtocolApplicationHandler.sol [hunk 6]
```diff
@@ -204,31 +207,90 @@ contract ProtocolApplicationHandler is
...
+                    fromScore,
+                    usdValueTotalWithrawals[_from],
+                    _transferValuation
+                );
+            }
         } else if (_action == ActionTypes.SELL ) {
-            if (isContract(_sender) && _to != _sender){ /// Non custodial sell 
+            if (_to != _sender){ /// Non custodial sell 
                 if (accountDenyForNoAccessLevel[ActionTypes.BUY].active) ruleProcessor.checkAccountDenyForNoAccessLevel(score);
+                if (accountMaxValueByAccessLevel[ActionTypes.BUY].active && _to != address(0))
```
Comment (oscarsernarosero) [inline src/client/application/ProtocolApplicationHandler.sol:248]: here

File: src/client/application/ProtocolApplicationHandler.sol [hunk 7]
```diff
@@ -204,31 +207,90 @@ contract ProtocolApplicationHandler is
...
+                    usdValueTotalWithrawals[_from] = ruleProcessor.checkAccountMaxValueOutByAccessLevel(
+                        accountMaxValueOutByAccessLevel[_action].ruleId,
+                        fromScore,
+                        usdValueTotalWithrawals[_from],
+                        _transferValuation
+                    );
+                }
             }
             if (accountDenyForNoAccessLevel[_action].active) ruleProcessor.checkAccountDenyForNoAccessLevel(fromScore);
+            if (accountMaxValueByAccessLevel[_action].active && _to != address(0))
```
Comment (oscarsernarosero) [inline src/client/application/ProtocolApplicationHandler.sol:260]: here

File: src/client/application/ProtocolApplicationHandler.sol [hunk 8]
```diff
@@ -204,31 +207,90 @@ contract ProtocolApplicationHandler is
...
+                usdValueTotalWithrawals[_from] = ruleProcessor.checkAccountMaxValueOutByAccessLevel(
+                    accountMaxValueOutByAccessLevel[_action].ruleId,
+                    fromScore,
+                    usdValueTotalWithrawals[_from],
+                    _transferValuation
+                );
+            }
         } else if (_action == ActionTypes.MINT) {
             if (accountDenyForNoAccessLevel[_action].active) ruleProcessor.checkAccountDenyForNoAccessLevel(score);
+            if (accountMaxValueByAccessLevel[_action].active && _to != address(0))
```
Comment (oscarsernarosero) [inline src/client/application/ProtocolApplicationHandler.sol:272]: here

File: src/client/application/ProtocolApplicationHandler.sol [hunk 9]
```diff
@@ -204,31 +207,90 @@ contract ProtocolApplicationHandler is
...
-        if (accountMaxValueByAccessLevel[_action].active && _to != address(0))
-            ruleProcessor.checkAccountMaxValueByAccessLevel(accountMaxValueByAccessLevel[_action].ruleId, score, _balanceValuation, _transferValuation);
-        if (accountMaxValueOutByAccessLevel[_action].active) {
-            usdValueTotalWithrawals[_from] = ruleProcessor.checkAccountMaxValueOutByAccessLevel(
-                accountMaxValueOutByAccessLevel[_action].ruleId,
-                fromScore,
-                usdValueTotalWithrawals[_from],
-                _transferValuation
-            );
+            if (accountMaxValueByAccessLevel[_action].active && _to != address(0))
```
Comment (oscarsernarosero) [inline src/client/application/ProtocolApplicationHandler.sol:284]: If this is inside the `if (_action == ActionTypes.BURN)` then it means that `_to = address(0)`. Not sure why we ask this here. Maybe Im not seeing a special case?

**Review by oscarsernarosero — APPROVED** [review]
Looks good. Im approving although I have a question about the `_to != address(0)` check. It seems unnecessary to me, but since the logic is just redundant and it seems to me like it doesn't really have any effect in the functionality of the code, Im approving. Im just curious of why of this check.

File: test/client/token/ERC721/marketplace/BuyerMarketplaceTests.t.sol [hunk 10]
```diff
@@ -1199,5 +1199,218 @@ contract MarketplaceNonCustodialTestsErc20SellsNftBuys is TokenUtils, ERC721Util
...
+
+        uint32 ruleId = createAccountMaxValueByRiskRule(riskScores, txLimits);
+        switchToAppAdministrator();
+        erc20Pricer.setSingleTokenPrice(address(applicationCoin), 10 * (10 ** 26)); //setting at $1
+        erc721Pricer.setNFTCollectionPrice(address(applicationNFTv2), buyPrice * (10 ** 26)); //setting at $1,000,000,000
+        vm.stopPrank();
+
+        setAccountMaxValueByRiskRuleSingleAction(ActionTypes.BUY, ruleId);
+
+        console.log(applicationHandler.isAccountMaxValueByRiskScoreActive(ActionTypes.BUY));
```
Comment (VoR0220) [inline test/client/token/ERC721/marketplace/BuyerMarketplaceTests.t.sol]: Probably need to get rid of console log here.

**Review by VoR0220 — APPROVED** [review]

**Pull request merged**